src/oc/storage/async/notification.clj

```diff
@@ -0,0 +1,142 @@
+(ns oc.storage.async.notification
+  "
+  Async publish of notification events to AWS SNS.
+  "
+  (:require [clojure.core.async :as async :refer (<!! >!!)]
+            [defun.core :refer (defun defun-)]
+            [taoensso.timbre :as timbre]
+            [cheshire.core :as json]
+            [amazonica.aws.sns :as sns]
+            [schema.core :as schema]
+            [oc.lib.schema :as lib-schema]
+            [oc.lib.time :as oc-time]
+            [oc.storage.config :as config]
+            [oc.storage.resources.common :as common-res]))
+
+;; ----- core.async -----
+
+(defonce notification-chan (async/chan 10000)) ; buffered channel
+
+(defonce notification-go (atom true))
+
+;; ----- Utility functions -----
+
+(defn- resource-type [content]
+  (cond
+    (:secure-uuid content) :entry
+    (:org-uuid content) :board
+    :else :org))
+
+;; ----- Data schema -----
+
+(defn notification-type? [notification-type]
+  (or (= notification-type :add) (= notification-type :update) (= notification-type :delete)))
+
+(defn resource-type? [resource-type]
+  (or (= resource-type :org) (= resource-type :board) (= resource-type :entry)))
+
+(def NotificationTrigger
+  "
+  A trigger for one of the various types of notifications that are published:
+
+  add - the content in  newly created, this happens when a board or entry is added
+  update - the content-id should be refreshed, this happens when a board or entry is updated
+  delete - the specified content-id is deleted, this happens when a board or entry is removed
+
+  The notification trigger contains the type of resource as `resource-type` and the current resource as `current`. If
+  the trigger is for an update it will also contain the updated resource as `update`. 
+
+  The user whose actions triggered the notification is included as `user`.
+
+  A timestamp for when the notice was created is included as `notice-at`.
+  "
+  {:notification-type (schema/pred notification-type?)
+   :resource-type (schema/pred resource-type?)
+   :current (schema/conditional #(= (resource-type %) :entry) common-res/Entry
```

Comment:
For `:current` and `:update` there is some code repeat using `schema/conditional`. Make this its own function similar to the `notification-type?` and `resource-type?` functions.

Reply:
We certainly could. `Schema/conditional` though is treating this code as data, and taking something like `common-res/Entry` to mean, validate that the data item matches the schema `common-res/Entry`. So if we refactored this into its own function, we'd gain some DRY but we'd lose some declarative. We'd need to call functions like `schema/validate` manually rather than declaratively. Neither is ideal, but I though the repetition was the better call here.

Reply:
Ok, seems fine as is after the clarification.